// @ng-dynamic-forms/core/esm2015/src/utils/autofill.utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
export const /** @type {?} */ AUTOCOMPLETE_OFF = "off";
export const /** @type {?} */ AUTOCOMPLETE_ON = "on";
export const /** @type {?} */ AUTOFILL_TOKEN_BILLING = "billing";
export const /** @type {?} */ AUTOFILL_TOKEN_SHIPPING = "shipping";
export const /** @type {?} */ AUTOFILL_TOKENS_ADDRESS = [AUTOFILL_TOKEN_BILLING, AUTOFILL_TOKEN_SHIPPING];
export const /** @type {?} */ AUTOFILL_TOKEN_HOME = "home";
export const /** @type {?} */ AUTOFILL_TOKEN_WORK = "work";
export const /** @type {?} */ AUTOFILL_TOKEN_MOBILE = "mobile";
export const /** @type {?} */ AUTOFILL_TOKEN_FAX = "fax";
export const /** @type {?} */ AUTOFILL_TOKEN_PAGER = "pager";
export const /** @type {?} */ AUTOFILL_TOKENS_CONTACT = [
    AUTOFILL_TOKEN_FAX, AUTOFILL_TOKEN_HOME, AUTOFILL_TOKEN_MOBILE, AUTOFILL_TOKEN_PAGER, AUTOFILL_TOKEN_WORK
];
export const /** @type {?} */ AUTOFILL_FIELD_STREET_ADDRESS = "street-address";
export const /** @type {?} */ AUTOFILL_FIELD_ADDRESS_LINE_1 = "address-line1";
export const /** @type {?} */ AUTOFILL_FIELD_ADDRESS_LINE_2 = "address-line2";
export const /** @type {?} */ AUTOFILL_FIELD_ADDRESS_LINE_3 = "address-line3";
export const /** @type {?} */ AUTOFILL_FIELD_ADDRESS_LEVEL_4 = "address-level4";
export const /** @type {?} */ AUTOFILL_FIELD_ADDRESS_LEVEL_3 = "address-level3";
export const /** @type {?} */ AUTOFILL_FIELD_ADDRESS_LEVEL_2 = "address-level2";
export const /** @type {?} */ AUTOFILL_FIELD_ADDRESS_LEVEL_1 = "address-level1";
export const /** @type {?} */ AUTOFILL_FIELD_NAME = "name";
export const /** @type {?} */ AUTOFILL_FIELD_HONORIFIC_PREFIX = "honorific-prefix";
export const /** @type {?} */ AUTOFILL_FIELD_GIVEN_NAME = "given-name";
export const /** @type {?} */ AUTOFILL_FIELD_ADDITIONAL_NAME = "additional-name";
export const /** @type {?} */ AUTOFILL_FIELD_FAMILY_NAME = "family-name";
export const /** @type {?} */ AUTOFILL_FIELD_HONORIFIC_SUFFIX = "honorific-suffix";
export const /** @type {?} */ AUTOFILL_FIELD_NICKNAME = "nickname";
export const /** @type {?} */ AUTOFILL_FIELD_USERNAME = "username";
export const /** @type {?} */ AUTOFILL_FIELD_NEW_PASSWORD = "new-password";
export const /** @type {?} */ AUTOFILL_FIELD_CURRENT_PASSWORD = "current-password";
export const /** @type {?} */ AUTOFILL_FIELD_ORGANIZATION_TITLE = "organization-title";
export const /** @type {?} */ AUTOFILL_FIELD_ORGANIZATION = "organization";
export const /** @type {?} */ AUTOFILL_FIELD_COUNTRY = "country";
export const /** @type {?} */ AUTOFILL_FIELD_COUNTRY_NAME = "country-name";
export const /** @type {?} */ AUTOFILL_FIELD_POSTAL_CODE = "postal-code";
export const /** @type {?} */ AUTOFILL_FIELD_CC_NAME = "cc-name";
export const /** @type {?} */ AUTOFILL_FIELD_CC_GIVEN_NAME = "cc-given-name";
export const /** @type {?} */ AUTOFILL_FIELD_CC_ADDITIONAL_NAME = "cc-additional-name";
export const /** @type {?} */ AUTOFILL_FIELD_CC_FAMILY_NAME = "cc-family-name";
export const /** @type {?} */ AUTOFILL_FIELD_CC_NUMBER = "cc-number";
export const /** @type {?} */ AUTOFILL_FIELD_CC_EXP = "cc-exp";
export const /** @type {?} */ AUTOFILL_FIELD_CC_EXP_MONTH = "cc-exp-month";
export const /** @type {?} */ AUTOFILL_FIELD_CC_EXP_YEAR = "cc-exp-year";
export const /** @type {?} */ AUTOFILL_FIELD_CC_CSC = "cc-csc";
export const /** @type {?} */ AUTOFILL_FIELD_CC_TYPE = "cc-type";
export const /** @type {?} */ AUTOFILL_FIELD_TRANSACTION_CURRENCY = "transaction-currency";
export const /** @type {?} */ AUTOFILL_FIELD_TRANSACTION_AMOUNT = "transaction-amount";
export const /** @type {?} */ AUTOFILL_FIELD_LANGUAGE = "language";
export const /** @type {?} */ AUTOFILL_FIELD_BDAY = "bday";
export const /** @type {?} */ AUTOFILL_FIELD_BDAY_DAY = "bday-day";
export const /** @type {?} */ AUTOFILL_FIELD_BDAY_MONTH = "bday-month";
export const /** @type {?} */ AUTOFILL_FIELD_BDAY_YEAR = "bday-year";
export const /** @type {?} */ AUTOFILL_FIELD_SEX = "sex";
export const /** @type {?} */ AUTOFILL_FIELD_URL = "url";
export const /** @type {?} */ AUTOFILL_FIELD_PHOTO = "photo";
export const /** @type {?} */ AUTOFILL_FIELDS = [
    AUTOFILL_FIELD_STREET_ADDRESS, AUTOFILL_FIELD_ADDRESS_LINE_1, AUTOFILL_FIELD_ADDRESS_LINE_2,
    AUTOFILL_FIELD_ADDRESS_LINE_3, AUTOFILL_FIELD_ADDRESS_LEVEL_4, AUTOFILL_FIELD_ADDRESS_LEVEL_3,
    AUTOFILL_FIELD_ADDRESS_LEVEL_2, AUTOFILL_FIELD_ADDRESS_LEVEL_1, AUTOFILL_FIELD_NAME,
    AUTOFILL_FIELD_HONORIFIC_PREFIX, AUTOFILL_FIELD_GIVEN_NAME, AUTOFILL_FIELD_ADDITIONAL_NAME,
    AUTOFILL_FIELD_FAMILY_NAME, AUTOFILL_FIELD_HONORIFIC_SUFFIX, AUTOFILL_FIELD_NICKNAME, AUTOFILL_FIELD_USERNAME,
    AUTOFILL_FIELD_NEW_PASSWORD, AUTOFILL_FIELD_CURRENT_PASSWORD, AUTOFILL_FIELD_ORGANIZATION_TITLE,
    AUTOFILL_FIELD_ORGANIZATION, AUTOFILL_FIELD_COUNTRY, AUTOFILL_FIELD_COUNTRY_NAME, AUTOFILL_FIELD_POSTAL_CODE,
    AUTOFILL_FIELD_CC_NAME, AUTOFILL_FIELD_CC_GIVEN_NAME, AUTOFILL_FIELD_CC_ADDITIONAL_NAME,
    AUTOFILL_FIELD_CC_FAMILY_NAME, AUTOFILL_FIELD_CC_NUMBER, AUTOFILL_FIELD_CC_EXP, AUTOFILL_FIELD_CC_EXP_MONTH,
    AUTOFILL_FIELD_CC_EXP_YEAR, AUTOFILL_FIELD_CC_CSC, AUTOFILL_FIELD_CC_TYPE, AUTOFILL_FIELD_TRANSACTION_CURRENCY,
    AUTOFILL_FIELD_TRANSACTION_AMOUNT, AUTOFILL_FIELD_LANGUAGE, AUTOFILL_FIELD_BDAY, AUTOFILL_FIELD_BDAY_DAY,
    AUTOFILL_FIELD_BDAY_MONTH, AUTOFILL_FIELD_BDAY_YEAR, AUTOFILL_FIELD_SEX, AUTOFILL_FIELD_URL, AUTOFILL_FIELD_PHOTO
];
export const /** @type {?} */ AUTOFILL_FIELD_TEL = "tel";
export const /** @type {?} */ AUTOFILL_FIELD_TEL_COUNTRY_CODE = "tel-country-code";
export const /** @type {?} */ AUTOFILL_FIELD_TEL_NATIONAL = "tel-national";
export const /** @type {?} */ AUTOFILL_FIELD_TEL_AREA_CODE = "tel-area-code";
export const /** @type {?} */ AUTOFILL_FIELD_TEL_LOCAL = "tel-local";
export const /** @type {?} */ AUTOFILL_FIELD_TEL_LOCAL_PREFIX = "tel-local-prefix";
export const /** @type {?} */ AUTOFILL_FIELD_TEL_LOCAL_SUFFIX = "tel-local-suffix";
export const /** @type {?} */ AUTOFILL_FIELD_TEL_LOCAL_EXTENSION = "tel-extension";
export const /** @type {?} */ AUTOFILL_FIELD_EMAIL = "email";
export const /** @type {?} */ AUTOFILL_FIELD_IMPP = "impp";
export const /** @type {?} */ AUTOFILL_FIELDS_CONTACT = [
    AUTOFILL_FIELD_TEL, AUTOFILL_FIELD_TEL_COUNTRY_CODE, AUTOFILL_FIELD_TEL_NATIONAL, AUTOFILL_FIELD_TEL_AREA_CODE,
    AUTOFILL_FIELD_TEL_LOCAL, AUTOFILL_FIELD_TEL_LOCAL_PREFIX, AUTOFILL_FIELD_TEL_LOCAL_SUFFIX,
    AUTOFILL_FIELD_TEL_LOCAL_EXTENSION, AUTOFILL_FIELD_EMAIL, AUTOFILL_FIELD_IMPP
];
export class AutoFillUtils {
    /**
     * @param {?} token
     * @return {?}
     */
    static isAddressToken(token) {
        return AUTOFILL_TOKENS_ADDRESS.indexOf(token) > -1;
    }
    /**
     * @param {?} token
     * @return {?}
     */
    static isContactField(token) {
        return AUTOFILL_FIELDS_CONTACT.indexOf(token) > -1;
    }
    /**
     * @param {?} token
     * @return {?}
     */
    static isContactToken(token) {
        return AUTOFILL_TOKENS_CONTACT.indexOf(token) > -1;
    }
    /**
     * @param {?} token
     * @return {?}
     */
    static isField(token) {
        return AUTOFILL_FIELDS.indexOf(token) > -1;
    }
    /**
     * @param {?} token
     * @return {?}
     */
    static isSectionToken(token) {
        return token.startsWith("section-");
    }
    /**
     * @param {?} tokens
     * @return {?}
     */
    static validate(tokens) {
        let /** @type {?} */ toExpression = (total, currentValue) => `${total}|${currentValue}`, /** @type {?} */
        tokensAddress = AUTOFILL_TOKENS_ADDRESS.reduce(toExpression), /** @type {?} */
        tokensContact = AUTOFILL_TOKENS_CONTACT.reduce(toExpression), /** @type {?} */
        fields = AUTOFILL_FIELDS.reduce(toExpression), /** @type {?} */
        fieldsContact = AUTOFILL_FIELDS_CONTACT.reduce(toExpression), /** @type {?} */
        regex = new RegExp(`^(section-\\w+\\s{1})?((${tokensAddress}){1}\\s)?((${fields}){1}|((${tokensContact}){1}\\s{1}(${fieldsContact})))$`);
        return regex.test(tokens);
    }
}
