// @ng-dynamic-forms/core/esm2015/src/component/dynamic-form-control.event.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/** @enum {string} */
const DynamicFormControlEventType = { Blur: "blur", Change: "change", Custom: "custom", Focus: "focus", };
export { DynamicFormControlEventType };
/**
 * @record
 */
export function DynamicFormControlEvent() { }
function DynamicFormControlEvent_tsickle_Closure_declarations() {
    /** @type {?} */
    DynamicFormControlEvent.prototype.$event;
    /** @type {?} */
    DynamicFormControlEvent.prototype.context;
    /** @type {?} */
    DynamicFormControlEvent.prototype.control;
    /** @type {?} */
    DynamicFormControlEvent.prototype.group;
    /** @type {?} */
    DynamicFormControlEvent.prototype.model;
    /** @type {?} */
    DynamicFormControlEvent.prototype.type;
}
/**
 * @record
 */
export function DynamicFormControlCustomEvent() { }
function DynamicFormControlCustomEvent_tsickle_Closure_declarations() {
    /** @type {?} */
    DynamicFormControlCustomEvent.prototype.customEvent;
    /** @type {?} */
    DynamicFormControlCustomEvent.prototype.customEventType;
}
/**
 * @param {?} $event
 * @return {?}
 */
export function isDynamicFormControlEvent($event) {
    return $event !== null && typeof $event === "object" && $event.hasOwnProperty("$event");
}
