// @ng-dynamic-forms/core/esm2015/src/component/dynamic-templateable-form-control.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { DynamicFormControlComponent } from "./dynamic-form-control.component";
/**
 * @abstract
 */
export class DynamicTemplateableFormControlComponent extends DynamicFormControlComponent {
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.layoutService
            .filterTemplatesByModel(this.model, this.templates)
            .forEach(template => this.bindTemplate(template));
    }
    /**
     * @param {?} template
     * @return {?}
     */
    bindTemplate(template) {
        if (typeof template.as === "string" && this.templateDirectives.has(template.as)) {
            let /** @type {?} */ property = /** @type {?} */ (this.templateDirectives.get(template.as));
            this.viewChild[property] = this.mapTemplate(template);
        }
    }
}
function DynamicTemplateableFormControlComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    DynamicTemplateableFormControlComponent.prototype.templateDirectives;
    /** @type {?} */
    DynamicTemplateableFormControlComponent.prototype.templates;
    /**
     * @abstract
     * @return {?}
     */
    DynamicTemplateableFormControlComponent.prototype.viewChild = function () { };
    /**
     * @abstract
     * @param {?} template
     * @return {?}
     */
    DynamicTemplateableFormControlComponent.prototype.mapTemplate = function (template) { };
}
