// @ng-dynamic-forms/core/esm2015/src/service/dynamic-form-layout.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Inject, Injectable, InjectionToken, Optional, QueryList } from "@angular/core";
import { DYNAMIC_TEMPLATE_DIRECTIVE_ALIGNMENT } from "../directive/dynamic-template.directive";
export const /** @type {?} */ DYNAMIC_FORM_CONTROL_MAP_FN = new InjectionToken("DYNAMIC_FORM_CONTROL_MAP_FN");
export class DynamicFormLayoutService {
    /**
     * @param {?} DYNAMIC_FORM_CONTROL_MAP_FN
     */
    constructor(DYNAMIC_FORM_CONTROL_MAP_FN) {
        this.DYNAMIC_FORM_CONTROL_MAP_FN = DYNAMIC_FORM_CONTROL_MAP_FN;
        this.DYNAMIC_FORM_CONTROL_MAP_FN = /** @type {?} */ (DYNAMIC_FORM_CONTROL_MAP_FN);
    }
    /**
     * @param {?} id
     * @param {?} formLayout
     * @return {?}
     */
    findById(id, formLayout) {
        if (formLayout !== null && typeof formLayout === "object") {
            for (let /** @type {?} */ key of Object.keys(formLayout)) {
                if (key === id) {
                    return formLayout[key];
                }
            }
        }
        return null;
    }
    /**
     * @param {?} model
     * @param {?} templates
     * @return {?}
     */
    filterTemplatesByModel(model, templates) {
        const /** @type {?} */ filterCallback = (template) => {
            return template.modelId === model.id || template.modelType === model.type;
        };
        if (templates instanceof QueryList) {
            return templates.filter(filterCallback);
        }
        else if (Array.isArray(templates)) {
            return templates.filter(filterCallback);
        }
        return [];
    }
    /**
     * @param {?} model
     * @param {?} templates
     * @param {?} alignment
     * @return {?}
     */
    getAlignedTemplate(model, templates, alignment) {
        return this.filterTemplatesByModel(model, templates)
            .find(template => template.as === null && template.align === alignment);
    }
    /**
     * @param {?} model
     * @param {?} templates
     * @return {?}
     */
    getStartTemplate(model, templates) {
        return this.getAlignedTemplate(model, templates, DYNAMIC_TEMPLATE_DIRECTIVE_ALIGNMENT.Start);
    }
    /**
     * @param {?} model
     * @param {?} templates
     * @return {?}
     */
    getEndTemplate(model, templates) {
        return this.getAlignedTemplate(model, templates, DYNAMIC_TEMPLATE_DIRECTIVE_ALIGNMENT.End);
    }
    /**
     * @param {?} layout
     * @param {?} context
     * @param {?} place
     * @return {?}
     */
    getClass(layout, context, place) {
        if (layout !== null && layout.hasOwnProperty(context)) {
            let /** @type {?} */ config = /** @type {?} */ (layout[context]);
            if (config.hasOwnProperty(place)) {
                return /** @type {?} */ (config[place]);
            }
        }
        return "";
    }
    /**
     * @param {?} model
     * @return {?}
     */
    getCustomComponentType(model) {
        if (this.DYNAMIC_FORM_CONTROL_MAP_FN) {
            return this.DYNAMIC_FORM_CONTROL_MAP_FN(model);
        }
        return null;
    }
}
DynamicFormLayoutService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
DynamicFormLayoutService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DYNAMIC_FORM_CONTROL_MAP_FN,] }, { type: Optional }] }
];
function DynamicFormLayoutService_tsickle_Closure_declarations() {
    /** @type {?} */
    DynamicFormLayoutService.prototype.DYNAMIC_FORM_CONTROL_MAP_FN;
}
